server/src/main/java/org/opensearch/indices/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,264 @@
+ * It also maintains a hash set of values which have had collisions. Values which haven't had collisions can be
+ * safely removed from the store. The fraction of collided values should be low,
+ * about 0.5% for a store with 10^7 values and a modulo of 2^28.
+ * The store estimates its memory footprint and will stop adding more values once it reaches its memory cap.
+ */
+public class RBMIntKeyLookupStore implements KeyLookupStore<Integer> {
+    protected final int modulo;
+    protected class KeyStoreStats {
+        protected int size;
+        protected long memSizeCapInBytes;
+        protected CounterMetric numAddAttempts;
+        protected CounterMetric numCollisions;
+        protected boolean guaranteesNoFalseNegatives;
+        protected int maxNumEntries;
+        protected boolean atCapacity;
+        protected CounterMetric numRemovalAttempts;
+        protected CounterMetric numSuccessfulRemovals;
+        protected KeyStoreStats(long memSizeCapInBytes, int maxNumEntries) {
+            this.size = 0;
+            this.numAddAttempts = new CounterMetric();
+            this.numCollisions = new CounterMetric();
+            this.memSizeCapInBytes = memSizeCapInBytes;
+            this.maxNumEntries = maxNumEntries;
+            this.atCapacity = false;
+            this.numRemovalAttempts = new CounterMetric();
+            this.numSuccessfulRemovals = new CounterMetric();
+        }
+    }
+
+    protected KeyStoreStats stats;
+    protected RoaringBitmap rbm;
+    private HashSet<Integer> collidedInts;
+    protected RBMSizeEstimator sizeEstimator;
+    protected final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    protected final Lock readLock = lock.readLock();
+    protected final Lock writeLock = lock.writeLock();
+
+    RBMIntKeyLookupStore(int modulo, long memSizeCapInBytes) {
+        this.modulo = modulo;
+        sizeEstimator = new RBMSizeEstimator(modulo);
+        this.stats = new KeyStoreStats(memSizeCapInBytes, calculateMaxNumEntries(memSizeCapInBytes));
+        this.rbm = new RoaringBitmap();
+        collidedInts = new HashSet<>();
+    }
+
+    protected int calculateMaxNumEntries(long memSizeCapInBytes) {
+        if (memSizeCapInBytes == 0) {
+            return Integer.MAX_VALUE;
+        }
+        return sizeEstimator.getNumEntriesFromSizeInBytes(memSizeCapInBytes);
+    }
+
+    protected final int transform(int value) {
+        return modulo == 0 ? value : value % modulo;
+    }
+
+    protected void handleCollisions(int transformedValue) {
+        stats.numCollisions.inc();
+        collidedInts.add(transformedValue);
+    }
+
+    @Override
+    public boolean add(Integer value) throws Exception {
+        if (value == null) {
+            return false;
+        }
+        writeLock.lock();
+        stats.numAddAttempts.inc();
+        try {
+            if (stats.size == stats.maxNumEntries) {
+                stats.atCapacity = true;
+                return false;
+            }
```

Comment:
We can move this check outside write lock?
For that we might have to change size, maxNumEntries to use thread safe counterMetric.